arch/arm/core/cortex_m/pm_s2ram.S

```diff
@@ -14,58 +14,138 @@
 #include <zephyr/arch/cpu.h>
 #include <zephyr/arch/common/pm_s2ram.h>
 
+/**
+ * Macro expanding to the (integer literal) offset of field
+ * `sr_name` in the __cpu_context structure. This macro has
+ * to be implemented in C because GEN_OFFSET_SYM provides the
+ * offsets as C preprocessor definitions; they are not available
+ * by the time the assembler macros are processed.
+ *
+ * See also: `arch/arm/core/offsets/offsets_aarch32.c`
+ */
+#define CPU_CTX_SR_OFFSET(sr_name) \
+	___cpu_context_t_ ## sr_name ## _OFFSET
+
+/**
+ * Macros used to save / load a special register in __cpu_context.
+ * These also have to be implemented in C due to CPU_CTX_SR_OFFSET.
+ */
+#define SAVE_SPECIAL_REG(sr_name, cpu_ctx_reg, tmp_reg)	\
+	mrs	tmp_reg, sr_name;			\
+	str	tmp_reg, [cpu_ctx_reg, # CPU_CTX_SR_OFFSET(sr_name)];
+
+#define RESTORE_SPECIAL_REG(sr_name, cpu_ctx_reg, tmp_reg)		\
+	ldr	tmp_reg, [cpu_ctx_reg, # CPU_CTX_SR_OFFSET(sr_name)];	\
+	msr	sr_name, tmp_reg;
+
 _ASM_FILE_PROLOGUE
 
 GTEXT(pm_s2ram_mark_set)
 GTEXT(pm_s2ram_mark_check_and_clear)
 GDATA(_cpu_context)
 
+# Pushes registers r4~r12 and lr on the stack.
+# r0 is unmodified, other GPRs may be overwritten.
```

Comment:
Nit: Let's try to stick with the comment style used throughout the Zephyr ASM files, which is C.